### PR TITLE
Pass callbacks to subplans during the subplan creation

### PR DESCRIPTION
**Pull request by adam-mazur-gat**
## Context
I got a report that wiping out users in our system takes waaaaay toooooooo loooong. And I put on my pants, took a magnifier and went out to the forest. 

I run the wipe out for the particular user and I immediately spot that we have too many SQL commands there.
I looked at them and it turned out that those commands are triggered by a code written in a plugin that we use for the plan. To be precise code was in the `before(:execution)` callback implemented by the plugin.

Let's go step by step. Here is the plan that was under the magnifier:

```ruby
class WipeOutAuditedPlugin
  include WipeOut::Plugin

  after(:execution) do |execution|
    record = execution.record
    
    Audited::Audit.create(auditable: record, action: "wipe_out",
      audited_changes: { "wipe_out_sensitive_personal_data" => true })
  end
end

CommentPlan = WipeOut.build_plan do
  wipe_out :content
  ignore :approved
end

UserPlan = WipeOut.build_plan do
  plugin WipeOutAuditedPlugin
  
  wipe_out :first_name, :last_name

  relation :comments, CommentPlan
end
```

and when I went `UserPlan.execute(user)` I spot that the number of audit records that were being added after every comment wipe out was increasing. 

1. After wiping out the first comment we got **one** `Audited::Audit` record
2. After wiping out the second comment we got `two` `Audited::Audit` records
...
3. After wiping out the 10th comment we got `ten` `Audited::Audit` records

Crazy, isn't it? 

I thought that it's something wrong with the user, but I tried another one and... the same.
I thought that it's something with the `WipeOutAuditedPlugin`, but it was ok.

So I went into this forbidden forest area called `wipe_out`. Old military terrain.

I found out that when the plan execution goes through the relations (association) it copies all callbacks from the parent plan to the relation plan - in our case it copies callbacks (provided by the audited plugin) from `UserPlan` to `CommentPlan`.

And here, putting my magnifier really close I spot a small **BUG** hiding under the leaf and here is how it looks:

https://github.com/GlobalAppTesting/wipe_out/blob/ea6f54838789a97eaf6548461ffbbee585083ff0/lib/wipe_out/execution/execute_plan.rb#L22-L31

Here for every record of the relation (every user comment we call `execute_on_record(plan, record)` (in our case `execute_on_record(CommentPlan, comment)` and as you can see the implementation

https://github.com/GlobalAppTesting/wipe_out/blob/ea6f54838789a97eaf6548461ffbbee585083ff0/lib/wipe_out/execution/execute_plan.rb#L42-L46

it creates a sub-execution like this:

https://github.com/GlobalAppTesting/wipe_out/blob/ea6f54838789a97eaf6548461ffbbee585083ff0/lib/wipe_out/execution/context.rb#L28-L31

and here spotlight on line 29:

https://github.com/GlobalAppTesting/wipe_out/blob/ea6f54838789a97eaf6548461ffbbee585083ff0/lib/wipe_out/execution/context.rb#L29

it copies callbacks to the relation plan - the problem is that it does it for **each relation record!!!**. For all relation records the same instance of relation plan is used and this is why we get more callbacks for every wiped association.

## Solution

I considered 2 solutions:
1. Change execution to use a new relation plan instance for every relation record. 
2. Move coping callbacks from the execution phase to plan-building phase. 

The first solution sounded to me like a bit of change and also not really an eco solution (1000 comments to wipe out = 1000 instances of the `CommentPlan`) - we are in the forest - zero waste.

Thus I went the second path. Callbacks are part of the plan and for the parent plan and e.g. included_plan they are set during plan-building, so I thought that setting them for the relation plan should also happen there.

## Changes 

- Move relation plan callbacks inheriting from execution to plan-building phase. I placed it at the end so plugins added in the plan will take precedence (to perceive the current behaviour).

### CODE REVIEW

File: spec/lib/wipe_out/plugins/logger_spec.rb
Comment (adam-mazur-gat) [inline spec/lib/wipe_out/plugins/logger_spec.rb:29]: Before I applied the change the updated spec was failing with 

```
[
    "[WipeOut] start plan=Plan(attributes=[:last_name, :access_tokens])",
    "[WipeOut] executing plan=Plan(attributes=[:last_name, :access_tokens]) record_class=User id=1",
    "[WipeOut] executing plan=Plan(attributes=[]) record_class=Comment id=1",
    "[WipeOut] wiped out plan=Plan(attributes=[]) record_class=Comment id=1",
    "[WipeOut] executing plan=Plan(attributes=[]) record_class=Comment id=2",
+   "[WipeOut] executing plan=Plan(attributes=[]) record_class=Comment id=2",
+   "[WipeOut] wiped out plan=Plan(attributes=[]) record_class=Comment id=2",
    "[WipeOut] wiped out plan=Plan(attributes=[]) record_class=Comment id=2",
    "[WipeOut] wiped out plan=Plan(attributes=[:last_name, :access_tokens]) record_class=User id=1",
    "[WipeOut] completed plan=Plan(attributes=[:last_name, :access_tokens])"
  ]
```

See the duplicated log lines for the second comment.

File: spec/lib/wipe_out/execute_spec.rb
Comment (adam-mazur-gat) [inline spec/lib/wipe_out/execute_spec.rb:204]: Result of the change in the factory.

File: README.md
Comment (adam-mazur-gat) [inline README.md:3]: CI was failing on `markdownlint`

image

**Comment by arathunku** [comment]
Great PR @adam-mazur-gat, thanks a lot for great description!

I belive we need to add another test into logger plugin, very similar to this one: https://github.com/GlobalAppTesting/wipe_out/blob/2041ea7a6f14c7e6541f651c7eb21d10e468fc71/spec/lib/wipe_out/execute_spec.rb#L181

It should use runtime selector to use plan. I'm worried that by moving callbacks to build phase this might get broken, and it looks like we don't have a test for that. It it's working too, then we're good to go :)

File: CHANGELOG.md
Comment (ktopolski) [inline CHANGELOG.md:5]: How about 
```suggestion
Fix issue with callbacks being executed multiple times in child plans
```

File: lib/wipe_out/plans/dsl.rb
Comment (ktopolski) [inline lib/wipe_out/plans/dsl.rb:75]: I can see it's in the if/else branch with condition on `plans` presence. Are relations with `plans` specified safe to use then?

File: spec/lib/wipe_out/plugins/logger_spec.rb
Comment (ktopolski) [reply spec/lib/wipe_out/plugins/logger_spec.rb:29]: wait, but what's the cause for that ? Isn't it an issue itself?

**Review by ktopolski — COMMENTED** [review]
Hey,

Thank you for this PR! I left some questions which I need an answer for before the approval :)

File: spec/lib/wipe_out/plugins/logger_spec.rb
Comment (adam-mazur-gat) [reply spec/lib/wipe_out/plugins/logger_spec.rb:29]: @ktopolski Hey - yes this is the bug that I found reproduced in the specs - red, green, refactor (here refactore wasn't needed)

the duplicated log lines for the second comment are present because for the second comment we already have 2 sets of callbacks.

**Comment by adam-mazur-gat** [comment]
@ktopolski @arathunku Very good spot about the `plans` case. Indeed they were broken after the fix - callbacks would not be passed in case of passing multiple plans to relation. Added a spec that failed and then applied a fix.

In the next PR I would like to:
- refactor `relation` method as it got quite long
- fix one spec that I found to be a false positive

File: spec/lib/wipe_out/plugins/logger_spec.rb
Comment (adam-mazur-gat) [inline spec/lib/wipe_out/plugins/logger_spec.rb:58]: I found that this is a false-positive (or I completely miss the point here) - I think it should be

 ```suggestion
        user_basic_plan.execute(create(:user))
```

to be fixed in another PR.

File: spec/lib/wipe_out/plugins/logger_spec.rb
Comment (adam-mazur-gat) [inline spec/lib/wipe_out/plugins/logger_spec.rb:91]: I wanted to have 2 of each type of comment to make sure that for each plan with multiple records it will work fine.

File: spec/lib/wipe_out/plugins/logger_spec.rb
Comment (ktopolski) [reply spec/lib/wipe_out/plugins/logger_spec.rb:58]: the example(`it`) description is about inheriting from top parent so imho yes, this should be user_basic_plan

Comment (adam-mazur-gat) [reply spec/lib/wipe_out/plugins/logger_spec.rb:58]: It didn't fix it here to keep this PR focused and also because after changing to `user_basic_plan` the spec fails :( and it's not related to the changes in this PR. @ktopolski

File: spec/lib/wipe_out/plugins/logger_spec.rb
Comment (ktopolski) [reply spec/lib/wipe_out/plugins/logger_spec.rb]: makes sense, thank you!

**Review by ktopolski — APPROVED** [review]
Great work with the `record.value.starts_with?("[SPECIAL]") ? vip_comments_plan : comments_plan` spec and fix :). With that covered we should be able to move forward with this PR. Thank you!

File: spec/lib/wipe_out/plugins/logger_spec.rb
Comment (arathunku) [reply spec/lib/wipe_out/plugins/logger_spec.rb:58]: +1, nice spot, should be `user_basic_plan`

**Review by arathunku — APPROVED** [review]
This looks great to me! I'd suggest 1 more MR -> submit this change to gat-app too and if all specs pass, good for merge and release. :) Miłosz is taking care of Tracker so that would cover 100% from our internal side.

**Comment by arathunku** [comment]
Thanks a lot @adam-mazur-gat for digging into this and fixing the issue.

**Pull request merged**
